From 5edcac1e04988b0c2e32a04c9726ecd394af981c Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 21 Feb 2018 22:43:52 +0000
Subject: [PATCH] TRITON-169 fwadm(1M) test runner should propagate nonzero
 program exits Reviewed by: Richard Kiene <richard.kiene@joyent.com> Approved
 by: Richard Kiene <richard.kiene@joyent.com>

---
 src/fw/test/runtest  | 10 +++++++++-
 src/fw/test/runtests | 21 ++++++++++++++++-----
 2 files changed, 25 insertions(+), 6 deletions(-)

diff --git a/src/fw/test/runtest b/src/fw/test/runtest
index 60727cd5..f63b7f74 100755
--- a/src/fw/test/runtest
+++ b/src/fw/test/runtest
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -o errexit
+
 UNAME=$(uname -s)
 NODE=node
 TOP=$(unset CDPATH; cd $(dirname $0)/../; pwd)
@@ -12,7 +14,7 @@ fi
 function usage
 {
     echo "Usage:"
-    echo "  runtest [OPTIONS...]"
+    echo "  runtest [OPTIONS...] file"
     echo ""
     echo "Options:"
     echo "  -r REPORTER  Nodeunit test reporter to use (default is 'tap')."
@@ -38,6 +40,12 @@ do
     esac
 done
 
+# Should have only one argument left.
+if [[ $OPTIND -ne $# ]]; then
+    usage
+    exit 1
+fi
+
 if [[ $UNAME == "SunOS" ]]; then
     echo "# Adding images.joyent.com as an imgadm source"
     imgadm sources -a "https://images.joyent.com/"
diff --git a/src/fw/test/runtests b/src/fw/test/runtests
index fb5e8143..bba381f7 100755
--- a/src/fw/test/runtests
+++ b/src/fw/test/runtests
@@ -88,6 +88,8 @@ mkdir -p $OUTPUT_DIR
 
 #---- start tests
 
+FINAL_EXIT_CODE=0
+
 for test in $(ls $TOP/test/integration | grep "\.test.js$"); do
     test_name=$(basename $test)
     set +o errexit
@@ -111,14 +113,21 @@ echo "# test output in $OUTPUT_DIR:"
 cd $OUTPUT_DIR
 ls *.tap
 
+if [[ ${test_errors} -gt 0 ]]; then
+    echo ""
+    echo "# failing tests:"
+    echo "$failed_tests"
+    FINAL_EXIT_CODE=1
+fi
+
 echo ""
 echo "# test results:"
 
 end_time=$(date +%s)
 elapsed=$((${end_time} - ${start_time}))
 
-tests=$(grep "^# tests [0-9]" $OUTPUT_DIR/*.tap | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
-passed=$(grep "^# pass  [0-9]" $OUTPUT_DIR/*.tap | tr -s ' ' | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
+tests=$(awk '/^# tests +[0-9]/ { sum += $(NF) } END { print sum }' $OUTPUT_DIR/*.tap)
+passed=$(awk '/^# pass +[0-9]/ { sum += $(NF) } END { print sum }' $OUTPUT_DIR/*.tap)
 [[ -z ${tests} ]] && tests=0
 [[ -z ${passed} ]] && passed=0
 fail=$((${tests} - ${passed}))
@@ -127,9 +136,11 @@ echo "# Completed in ${elapsed} seconds."
 echo -e "# \033[32mPASS: ${passed} / ${tests}\033[39m"
 if [[ ${fail} -gt 0 ]]; then
     echo -e "# \033[31mFAIL: ${fail} / ${tests}\033[39m"
+    FINAL_EXIT_CODE=1
+elif [[ ${test_errors} -gt 0 ]]; then
+    echo -e "# \033[31mFAIL: ? / ${tests}\033[39m"
+    FINAL_EXIT_CODE=1
 fi
 echo ""
 
-if [[ ${tests} != ${passed} ]]; then
-    exit 1
-fi
+exit $FINAL_EXIT_CODE
-- 
2.21.0

